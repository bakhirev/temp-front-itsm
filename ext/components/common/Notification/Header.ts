import styled from 'styled-components';

import { Subtitle3 } from '../../Typography';

export const Header = styled(Subtitle3)<{ hasOffset?: boolean }>`
  ${({ hasOffset }) => (hasOffset ? 'margin-bottom: 4px;' : '')}
`;
