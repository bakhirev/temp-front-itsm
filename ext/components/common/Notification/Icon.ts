import styled from 'styled-components';

export const Icon = styled.div`
  position: absolute;
  left: 12px;
  top: 12px;
`;
