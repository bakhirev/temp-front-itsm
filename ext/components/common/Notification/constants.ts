export const TRANSFORM_SCALE = 0.9;
export const DEFAULT_DURATION = 500;

export const ICON_CLOSE_RIGHT = 16;
export const ICON_CLOSE_TOP = 12;

export const STATIC_PADDING = '12px 16px 12px 52px';
export const DYNAMIC_PADDING = '14px 52px';
export const MOBILE_DYNAMIC_PADDING = '14px 24px';

export const MOBILE_MARGIN = '16px';
export const TOP_MARGIN = '16px';
