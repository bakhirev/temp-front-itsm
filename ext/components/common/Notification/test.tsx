import React, { FC } from 'react';
import { shallow } from 'enzyme';

import type { INotificationProps } from './NotificationComponent';
import type { IDynamicNotificationProps } from './DynamicNotification';

export const runCommonTests = (Notification: FC<INotificationProps>, caption: string) => {
  const header = caption + ' Header';
  const message = caption + ' Message';
  return [
    it('should render component by default as "INFO" notification without header', () => {
      const wrapper = shallow(<Notification>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component by default as "INFO" notification with header', () => {
      const wrapper = shallow(<Notification header={header}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="INFO" notification without header', () => {
      const wrapper = shallow(<Notification kind={'info'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="INFO" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'info'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="WARNING" notification without header', () => {
      const wrapper = shallow(<Notification kind={'warning'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="WARNING" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'warning'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="SUCCESS" notification without header', () => {
      const wrapper = shallow(<Notification kind={'success'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="SUCCESS" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'success'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="ERROR" notification without header', () => {
      const wrapper = shallow(<Notification kind={'error'}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component with kind="ERROR" notification with header', () => {
      const wrapper = shallow(
        <Notification kind={'error'} header={header}>
          {message}
        </Notification>
      );
      expect(wrapper).toMatchSnapshot();
    }),

    it('should render component by default as "INFO" notification without header with custom width', () => {
      const wrapper = shallow(<Notification width={777}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),
  ];
};

export const runCommonDynamicTests = (
  Notification: FC<IDynamicNotificationProps>,
  caption: string
) => {
  const message = caption + ' Message';
  return [
    it('should render component inside the document.body container', () => {
      const wrapper = shallow(<Notification container={document.body}>{message}</Notification>);
      expect(wrapper).toMatchSnapshot();
    }),
  ];
};
