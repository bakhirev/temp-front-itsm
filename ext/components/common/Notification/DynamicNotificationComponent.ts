import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../default-theme';
import { MEDIA_MOBILE } from '../media';
import { Z_INDEXES } from '../z-indexes';

import { NotificationComponent } from './NotificationComponent';
import { Header } from './Header';
import { Message } from './Message';
import { Icon } from './Icon';
import { IconClose } from './IconClose';
import { getInverseColor } from './utils';
import { DYNAMIC_PADDING, MOBILE_DYNAMIC_PADDING, MOBILE_MARGIN, TOP_MARGIN } from './constants';
import type { INotificationComponentProps } from './NotificationComponent';

export interface IDynamicNotificationComponentProps extends INotificationComponentProps {
  inverse?: boolean;
  mobile?: boolean;
  managed?: boolean;
  top?: number;
  left?: number;
  right?: number;
}

export const DynamicNotificationComponent = styled(
  NotificationComponent
)<IDynamicNotificationComponentProps>`
  padding: ${DYNAMIC_PADDING};

  ${({ top, left, right, managed }) =>
    managed
      ? `
    ${top !== undefined ? `margin-top: ${top}px;` : ''}
    ${left !== undefined ? `margin-left: ${left}px;` : ''}
    ${
      right !== undefined
        ? `
      margin-left: auto;
      margin-right: ${right}px;
    `
        : ''
    }
    &:not(:first-child) {
      margin-top: ${TOP_MARGIN};
    }
    @media ${MEDIA_MOBILE} {
      margin-top: ${MOBILE_MARGIN};
      margin-left: ${MOBILE_MARGIN};
    }
  `
      : `
    position: fixed;
    z-index: ${Z_INDEXES.NOTIFICATION};
    ${top !== undefined ? `top: ${top}px;` : ''}
    ${left !== undefined ? `left: ${left}px;` : ''}
    ${right !== undefined ? `right: ${right}px;` : ''}
    &:not(:first-child) {
      top: ${TOP_MARGIN};
    }
    @media ${MEDIA_MOBILE} {
      top: ${MOBILE_MARGIN};
      left: ${MOBILE_MARGIN};
    }
  `}

  @media ${MEDIA_MOBILE} {
    width: calc(100% - 2 * ${MOBILE_MARGIN});
    padding: ${MOBILE_DYNAMIC_PADDING};

    ${Icon} {
      display: none;
    }

    ${({ mobile }) =>
      mobile &&
      css`
        ${IconClose} {
          display: none;
        }
      `}
  }

  ${({ theme, kind, inverse }) =>
    inverse
      ? `
        background-color: ${theme.color.neutral[80]};
        border-left: 4px solid ${getInverseColor(kind, theme)};
        ${Header}, ${Message} {
          color: ${theme.color.neutral.white};
        }
        ${Icon} svg {
          fill: ${getInverseColor(kind, theme)};
        }
        ${IconClose} svg {
          fill: ${theme.color.neutral.white};
          &:hover {
            fill: ${theme.color.primary[40]};
          }
        }`
      : `
        box-shadow: ${theme.boxShadow[4]};
        ${IconClose} svg {
          fill: ${theme.color.neutral[50]};
          &:hover {
            fill: ${theme.color.primary[60]};
          }
        }
  `}
`;

DynamicNotificationComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
