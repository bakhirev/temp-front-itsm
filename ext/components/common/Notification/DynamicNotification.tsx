import React, { useState, useEffect, FC } from 'react';
import { Transition } from 'react-transition-group';
import ReactDOM from 'react-dom';

import { ReactComponent as CloseOutline } from '../../Icons/service/CloseOutline.svg';
import { useGesture } from '../hooks';
import { useMobile } from '../hooks/useMobile';

import type { INotificationProps } from './NotificationComponent';
import { NOTIFICATION_ICONS } from './NotificationComponent';
import { DynamicNotificationComponent } from './DynamicNotificationComponent';
import { Icon } from './Icon';
import { IconClose } from './IconClose';
import { DEFAULT_DURATION } from './constants';
import { getTransitionStyles } from './utils';

export type VisibilityHandler = () => void;

export interface IDynamicNotificationProps extends INotificationProps {
  /** Время отображения анимации открытия/закрытия */
  duration?: number;
  /** Время отображения оповещения до автоматического закрытия */
  delay?: number;
  /** Контейнер, в котором происходит размещение оповещения */
  container?: Element;
  /** Коллбэк начала анимации отображения оповещения */
  onShow?: VisibilityHandler;
  /** Коллбэк окончания анимации скрытия оповещения */
  onHide?: VisibilityHandler;
}

const DEFAULT_WIDTH = '100%';

interface IDynamicNotification extends IDynamicNotificationProps {
  inverse?: boolean;
  top?: number;
  left?: number;
  right?: number;
}

export const DynamicNotification: FC<IDynamicNotification> = ({
  children,
  kind = 'info',
  width = DEFAULT_WIDTH,
  duration = DEFAULT_DURATION,
  delay,
  container,
  header,
  inverse,
  top,
  left,
  right,
  className,
  dataTestId,
  onShow,
  onHide,
  ...props
}) => {
  const KindIcon = NOTIFICATION_ICONS[kind];
  const mobile = useMobile();
  const [gesture, gestureBinding] = useGesture();
  const [visible, setVisible] = useState(true);
  const [isMounted, setIsMounted] = useState(false);
  useEffect(() => {
    setIsMounted(true);
  }, []);

  useEffect(() => {
    (gesture === 'left' || gesture === 'right') && setVisible(false);
  }, [gesture, setVisible]);

  const applyTimer = () => {
    if (delay) {
      const timer = setTimeout(() => {
        setVisible(false);
        clearTimeout(timer);
      }, delay);
    }
  };

  const transitionStyles = getTransitionStyles(duration);
  const component = (
    <Transition
      in={isMounted && visible}
      timeout={duration}
      onEntered={() => applyTimer()}
      onEnter={() => onShow && onShow()}
      onExited={() => onHide && onHide()}
    >
      {(state) => (
        <DynamicNotificationComponent
          {...gestureBinding}
          {...props}
          mobile={mobile}
          style={{ ...transitionStyles[state] }}
          className={className}
          data-test-id={dataTestId}
          kind={kind}
          width={width}
          inverse={inverse}
          top={top}
          left={left}
          right={right}
        >
          <Icon>
            <KindIcon />
          </Icon>
          <IconClose onClick={() => setVisible(false)} inverse={inverse} markerOffset={6}>
            <CloseOutline width={20} height={20} />
          </IconClose>
          {children}
        </DynamicNotificationComponent>
      )}
    </Transition>
  );
  return container ? ReactDOM.createPortal(component, container) : component;
};
