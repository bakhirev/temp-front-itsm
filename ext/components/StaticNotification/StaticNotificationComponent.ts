import styled from 'styled-components';

import { DEFAULT_THEME } from '../common';
import { NotificationComponent } from '../common/Notification/NotificationComponent';
import type { INotificationComponentProps } from '../common/Notification/NotificationComponent';

interface IStaticNotificationComponentProps extends INotificationComponentProps {
  showWhiteBackground?: boolean;
}

export const StaticNotificationComponent = styled(
  NotificationComponent
)<IStaticNotificationComponentProps>`
  ${({ theme, showWhiteBackground }) => `
     ${showWhiteBackground ? `background-color: ${theme.color.neutral.white}` : ''};
  `}
`;

StaticNotificationComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
