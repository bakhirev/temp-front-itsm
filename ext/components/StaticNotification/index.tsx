import React, { FC } from 'react';

import {
  INotificationProps,
  NOTIFICATION_ICONS,
} from '../common/Notification/NotificationComponent';
import { Icon } from '../common/Notification/Icon';
import { Header } from '../common/Notification/Header';
import { Message } from '../common/Notification/Message';

import { StaticNotificationComponent } from './StaticNotificationComponent';

export interface IStaticNotificationProps extends INotificationProps {
  /** Принудительно белый фон (необходим при нахождении в контейнере с серым фоном) */
  showWhiteBackground?: boolean;
}

const DEFAULT_WIDTH = '320px';

export const StaticNotification: FC<IStaticNotificationProps> = ({
  children,
  kind = 'info',
  width = DEFAULT_WIDTH,
  header,
  showWhiteBackground,
  className,
  dataTestId,
}: IStaticNotificationProps) => {
  const KindIcon = NOTIFICATION_ICONS[kind];
  return (
    <StaticNotificationComponent
      className={className}
      data-test-id={dataTestId}
      kind={kind}
      width={width}
      showWhiteBackground={showWhiteBackground}
    >
      <Icon>
        <KindIcon />
      </Icon>
      {!!header && <Header hasOffset>{header}</Header>}
      <Message>{children}</Message>
    </StaticNotificationComponent>
  );
};
