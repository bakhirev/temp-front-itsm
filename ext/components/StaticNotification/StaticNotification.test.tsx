import React from 'react';
import { shallow } from 'enzyme';

import { runCommonTests } from '../common/Notification/test';
import { StaticNotification } from '../StaticNotification';

const message = 'StaticNotification Message';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

describe('StaticNotification', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  runCommonTests(StaticNotification, 'StaticNotification');

  it('should render component by default as "INFO" notification with white background', () => {
    const wrapper = shallow(<StaticNotification showWhiteBackground>{message}</StaticNotification>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind="ERROR" notification with white background', () => {
    const wrapper = shallow(
      <StaticNotification kind={'error'} showWhiteBackground>
        {message}
      </StaticNotification>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
