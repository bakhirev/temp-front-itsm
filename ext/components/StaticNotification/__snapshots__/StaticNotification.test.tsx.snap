// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaticNotification should render component by default as "INFO" notification with header 1`] = `
<Styled(styled.div)
  kind="info"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)
    hasOffset={true}
  >
    StaticNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component by default as "INFO" notification with white background 1`] = `
<Styled(styled.div)
  kind="info"
  showWhiteBackground={true}
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component by default as "INFO" notification without header 1`] = `
<Styled(styled.div)
  kind="info"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component by default as "INFO" notification without header with custom width 1`] = `
<Styled(styled.div)
  kind="info"
  theme={Object {}}
  width={777}
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="ERROR" notification with header 1`] = `
<Styled(styled.div)
  kind="error"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)
    hasOffset={true}
  >
    StaticNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="ERROR" notification with white background 1`] = `
<Styled(styled.div)
  kind="error"
  showWhiteBackground={true}
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="ERROR" notification without header 1`] = `
<Styled(styled.div)
  kind="error"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="INFO" notification with header 1`] = `
<Styled(styled.div)
  kind="info"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)
    hasOffset={true}
  >
    StaticNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="INFO" notification without header 1`] = `
<Styled(styled.div)
  kind="info"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="SUCCESS" notification with header 1`] = `
<Styled(styled.div)
  kind="success"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)
    hasOffset={true}
  >
    StaticNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="SUCCESS" notification without header 1`] = `
<Styled(styled.div)
  kind="success"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="WARNING" notification with header 1`] = `
<Styled(styled.div)
  kind="warning"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)
    hasOffset={true}
  >
    StaticNotification Header
  </Styled(styled.div)>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`StaticNotification should render component with kind="WARNING" notification without header 1`] = `
<Styled(styled.div)
  kind="warning"
  theme={Object {}}
  width="320px"
>
  <styled.div>
    <div />
  </styled.div>
  <Styled(styled.div)>
    StaticNotification Message
  </Styled(styled.div)>
</Styled(styled.div)>
`;
